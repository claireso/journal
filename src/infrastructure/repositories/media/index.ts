export { default as MediaRepositoryImpl } from './MediaRepositoryImpl'
