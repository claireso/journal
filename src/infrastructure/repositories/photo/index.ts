export { default as PhotoRepositoryImpl } from './PhotoRepositoryImpl'
