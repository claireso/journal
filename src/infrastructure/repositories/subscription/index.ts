export { default as SubscriptionRepositoryImpl } from './SubscriptionRepositoryImpl'
