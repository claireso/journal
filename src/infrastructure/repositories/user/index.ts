export { default as UserRepositoryImpl } from './UserRepositoryImpl'
