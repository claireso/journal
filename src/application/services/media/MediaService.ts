import path from 'path'
import { unlink } from 'fs/promises'
import { MediaType } from '@domain/entities'
import { NotFoundError } from '@domain/errors'
import { MediaRepository } from '@domain/repositories'
import uploadFile from '@utils/uploadFile'

export default class MediaService {
  private repository: MediaRepository
  private logger: unknown

  constructor(repository: MediaRepository, logger: unknown) {
    this.repository = repository
    this.logger = logger
  }

  async create(file: File) {
    const { filename, width, height } = await uploadFile(file)

    const data = {
      type: MediaType.IMAGE,
      name: filename,
      size: {
        width: width,
        height: height
      }
    }

    return this.repository.create(data)
  }

  async getById(id: number) {
    const media = await this.repository.getById(id)

    if (media === null) {
      throw new NotFoundError(`Not found media`, { cause: { photoId: id } })
    }

    return media
  }

  async delete(id: number) {
    const media = await this.getById(id)

    await this.repository.delete(id)

    // delete media file from the folder
    await unlink(path.resolve('uploads', media.name))
  }
}
