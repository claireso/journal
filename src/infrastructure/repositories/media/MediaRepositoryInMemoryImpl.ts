import { Media, MediaType, mapRowToMedia } from '@domain/entities'
import { MediaRepository } from '@domain/repositories'

export default class MediaRepositoryInMemoryImpl implements MediaRepository {
  private media: Media[]

  constructor() {
    this.media = [
      {
        id: 1,
        type: MediaType.IMAGE,
        name: 'abc.jpg',
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        size: {
          width: 1640,
          height: 1093
        }
      },
      {
        id: 2,
        type: MediaType.IMAGE,
        name: 'def.jpg',
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        size: {
          width: 800,
          height: 1200
        }
      },
      {
        id: 3,
        type: MediaType.IMAGE,
        name: 'ghi.jpg',
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        size: {
          width: 800,
          height: 800
        }
      },
      {
        id: 4,
        type: MediaType.IMAGE,
        name: 'jkl.jpg',
        created_at: new Date('2024-10-05T17:12:40.400Z'),
        size: {
          width: 1640,
          height: 1093
        }
      }
    ]
  }

  async create(data: Pick<Media, 'type' | 'name' | 'size'>): Promise<Media> {
    const {
      type,
      name,
      size: { width, height }
    } = data

    const lastMedia = this.media.at(-1) as Media
    const newMedia = mapRowToMedia({
      id: lastMedia.id + 1,
      created_at: new Date('2024-10-05T17:12:40.400Z'),
      type,
      name,
      width,
      height
    })
    this.media.push(newMedia)
    return newMedia
  }

  async getById(id: number): Promise<Media | null> {
    const media = this.media.find((_media) => _media.id === id)
    return media || null
  }

  async delete(id: number): Promise<void> {
    this.media = this.media.filter((_media) => _media.id !== id)
  }
}
