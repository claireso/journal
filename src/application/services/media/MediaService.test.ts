import fs from 'fs'
// import { Jimp } from 'jimp'
import sharp from 'sharp'
import { MediaRepository } from '@domain/repositories'
import MediaRepositoryInMemoryImpl from '@infrastructure/repositories/media/MediaRepositoryInMemoryImpl'
import MediaService from './MediaService'

describe('application/MediaService', () => {
  let mediaRepository: MediaRepository
  let mediaService: MediaService

  const createFakeImageFile = async () => {
    const buffer = await sharp({
      create: { width: 3, height: 3, channels: 4, background: { r: 255, g: 0, b: 0 } }
    })
      .jpeg()
      .toBuffer()
    const blob = new Blob([buffer], { type: 'image/jpeg' })
    const file = new File([blob], 'test.jpg', { type: 'image/jpeg' })

    return file
  }

  beforeEach(() => {
    mediaRepository = new MediaRepositoryInMemoryImpl()
    mediaService = new MediaService(mediaRepository, console)
  })

  describe('CREATE', () => {
    it('should create a media successfully', async () => {
      // arrange
      const file = await createFakeImageFile()

      // act
      const media = await mediaService.create(file)

      // assert
      expect(media).toMatchSnapshot({
        name: expect.any(String)
      })
    })

    it('should throw an error when creating a media with invalid file', async () => {
      // arrange
      const file = new File(['Invalid data!'], 'test.jpg', { type: 'invalid/data' })

      // act
      const promise = mediaService.create(file)

      // act + assert
      await expect(promise).rejects.toThrow()
    })
  })

  describe('GET', () => {
    it('should get a media by identifier', async () => {
      // act
      const media = await mediaService.getById(1)

      // assert
      expect(media).toMatchSnapshot()
    })

    it('should throw a NotFoundError error when getting a media with invalid id', async () => {
      // act
      const promise = mediaService.getById(1000)

      // arrange
      await expect(promise).rejects.toThrow('Not found media')
    })
  })

  describe('DELETE', () => {
    it('should delete media successfully', async () => {
      // arrange
      const file = await createFakeImageFile()
      const media = await mediaService.create(file)

      // act
      await mediaService.delete(media.id)

      // assert
      expect(fs.existsSync(`./uploads/${media.name}`)).toBe(false)
    })

    it('should throw a NotFoundError error when deleting a media with invalid id', async () => {
      // act
      const promise = mediaService.delete(1000)
      // assert
      expect(promise).rejects.toThrow('Not found media')
    })
  })
})
