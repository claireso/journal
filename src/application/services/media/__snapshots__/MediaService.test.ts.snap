// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application/MediaService CREATE should create a media successfully 1`] = `
{
  "created_at": 2024-10-05T17:12:40.400Z,
  "id": 5,
  "name": Any<String>,
  "size": {
    "height": 3,
    "width": 3,
  },
  "type": "image",
}
`;

exports[`application/MediaService GET should get a media by identifier 1`] = `
{
  "created_at": 2024-10-05T17:12:40.400Z,
  "id": 1,
  "name": "abc.jpg",
  "size": {
    "height": 1093,
    "width": 1640,
  },
  "type": "image",
}
`;
